hie-plugin-api/Haskell/Ide/Engine/GhcModuleCache.hs

```diff
@@ -100,21 +102,59 @@ data LookupCradleResult = ReuseCradle | LoadCradle CachedCradle | NewCradle File
 lookupCradle :: FilePath -> GhcModuleCache -> LookupCradleResult
 lookupCradle fp gmc =
   case currentCradle gmc of
-    Just (dirs, _c) | (any (\d -> d `isPrefixOf` fp) dirs) -> ReuseCradle
+    Just (dirs, _c, _) | any (`isPrefixOf` fp) dirs -> ReuseCradle
     _ -> case T.match (cradleCache gmc) (B.pack fp) of
            Just (_k, c, _suf) -> LoadCradle c
            Nothing  -> NewCradle fp
 
-data CachedCradle = CachedCradle BIOS.Cradle HscEnv
+getComponentOptions
+  :: (MonadIO m, HasGhcModuleCache m)
+  => FilePath
+  -> m (Maybe BIOS.ComponentOptions)
+getComponentOptions fp = do
+  mc <- getModuleCache
+  case currentCradle mc of
+    Just (dirs, cradle, co) | any (`isPrefixOf` fp) dirs -> case co of
+      Just _ -> return co
+      _      -> setCo fp cradle setCur
+    _ -> case T.match (cradleCache mc) (B.pack fp) of
+      Just (_, CachedCradle _ _ (Just co), _) -> return $ Just co
+      Just (p, CachedCradle cradle _ Nothing, _) ->
+        setCo (B.unpack p) cradle setCac
+      _ -> return Nothing
+ where
+  setCo fp' cradle mod' = do
+    res <- liftIO $ BIOS.getCompilerOptions fp' cradle
```

Comment:
I dont really like this. You potentially block a thread for an undetermined amount of time (potentially for minutes) and the user gets zero feedback, just wondering why nothing works. Also, on which thread gets this evaluated? I think it is the `Scheduler` thread, but I am not sure. If it is the `IDE` thread, then it is truly not acceptable as we require that nothing on the `IDE` thread has a blocking call.
 I would prefer if that function acts more like `lookupCradle` that it only looks up the component options if they are already cached, e.g. `lookupComponentOptions`.

Reply:
However, this might have other impliciations, e.g. what happens when you try to format a file before it has loaded? I suppose that should not be allowed?

Reply:
Blocking is limited since, the cradle needs to be initialized before we attempt `BIOS.getCompilerOptions`. 

If the cradle is not loaded we may format the document incorrectly. While we could fail to format before the cradle is loaded the current behavior allows you to format even when hie-bios is unable to load the cradle. It's up to the user to decided if they can format without the cradle options. I still get strange `unable to satisfy z-package`  errors when I load a project from a particular broken states, so I don't want my formatter to be depend on having a loaded cradle.

Reply:
> Blocking is limited since, the cradle needs to be initialized before we attempt BIOS.getCompilerOptions.

When you use `hie.yaml`, the function `BIOS.getCompilerOptions` will compile all dependencies of the project if they havent been built before. So, it might block for multiple minutes.

> It's up to the user to decided if they can format without the cradle options. I still get strange unable to satisfy z-package errors when I load a project from a particular broken states, so I don't want my formatter to be depend on having a loaded cradle.

When this happens, the compiler options may be just as broken. In general the problem is that formatting depends very much on compiler options that brittany and ormolu and floskell can not obtain by themselves. I think, failing to format if we cant load the project is the more expected behaviour, since we can only format the module correctly if it compiles.

> I still get strange unable to satisfy z-package errors 

This happens because `stack repl` fails when you have internal libraries. When you use cabal, this is not a problem, iirc.

Reply:
Can a cradle end up in `GhcModuleCache` prior to it's initialization? My thinking was since `getCompilerOptions` was already called in initialization, the second call will not have to build dependencies, except in pathological cases.

"splitting the call to `BIOS.initializeFlagsWithCradleWithMessage` into its implementation" will replace this any way.

Reply:
You are right that should not happen. But it may if someone changes the configuration in between calls, e.g. modifies `package.yaml` or `*.cabal`.

Reply:
> When this happens, the compiler options may be just as broken. In general the problem is that formatting depends very much on compiler options that brittany and ormolu and floskell can not obtain by themselves. I think, failing to format if we cant load the project is the more expected behaviour, since we can only format the module correctly if it compiles.

Each formatter can handle this differently. Ormolu always enables all but a few extensions, so most of the time it's behavior is correct, the user can always just undo it, if it's incorrect. Working without a valid cradle especially after I enable ranged formats for Ormolu, is a behavior I want to preserve.